### ISSUE TITLE
Karpenter couldn't create new nodes and showing the LaunchTemplate not found error

### ISSUE THREAD

**Issue by nandiheath** [aws/karpenter#1964]
## Version

**Karpenter**: v0.10.1

**Kubernetes**: v1.21.12

## Expected Behavior
Karpenter should be able to create new nodes when there is unscheduled pods pending

## Actual Behavior
Karpenter showed error when trying to create new nodes for unscheduled pods:
```
ERROR	controller	Launching node, creating cloud provider machine, creating fleet InvalidLaunchTemplateName.NotFoundException: The specified launch template, with template name Karpenter-eks-us-west-2-7225536031179167781, does not exist.
```

## Steps to Reproduce the Problem
We are running HA Karpenter with 3 pods, that has provisioner's node TTL set to 24 hours, and we are not using any custom Launch Template i.e. Karpenter should manage its own LT which picks up the latest AMI image.

We enabled node TTL last week and it was running fine until yesterday ET 18:00, which Karpenter no longer could able spawn new nodes because of the error above. The issue is figured and resolved by a hard restarting of the leader pod.

## Resource Specs and Logs
I removed the noise but please let me know if other logs could help:

```
# datetime, pod_name, time, level, content
Jun 20 18:06:39.646 | karpenter-65456c857f-ztqxq | 2022-06-20T22:06:39.554Z ERROR controller Launching node, creating cloud provider machine, creating fleet InvalidLaunchTemplateName.NotFoundException: The specified launch template, with template name Karpenter-eks-us-west-2-7225536031179167781, does not exist.
-- | -- | --
  | Jun 20 18:06:34.637 | karpenter-65456c857f-ztqxq | 2022-06-20T22:06:34.485Z ERROR controller Launching node, creating cloud provider machine, creating fleet InvalidLaunchTemplateName.NotFoundException: The specified launch template, with template name Karpenter-eks-us-west-2-7225536031179167781, does not exist.
  | Jun 20 18:06:32.351 | karpenter-65456c857f-ct9dl | 2022-06-20T22:06:31.789Z DEBUG controller.aws.launchtemplate Deleted launch template lt-0de5cf95221712a63 {"commit": "10fc37b"}
  | Jun 20 18:06:00.631 | karpenter-65456c857f-ztqxq | 2022-06-20T22:06:00.618Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 18:03:45.630 | karpenter-65456c857f-ztqxq | 2022-06-20T22:03:45.510Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 18:01:51.632 | karpenter-65456c857f-ztqxq | 2022-06-20T22:01:51.049Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 17:59:55.637 | karpenter-65456c857f-ztqxq | 2022-06-20T21:59:55.110Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 17:58:03.637 | karpenter-65456c857f-ztqxq | 2022-06-20T21:58:02.980Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 17:57:08.117 | karpenter-65456c857f-ztqxq | 2022-06-20T21:56:38.814Z DEBUG controller.events Normal {"commit": "10fc37b", "object": {"kind":"Lease","namespace":"karpenter","name":"karpenter-leader-election","uid":"b92de02a-24a4-4870-bf3d-25e00300469e","apiVersion":"coordination.k8s.io/v1","resourceVersion":"3322701528"}, "reason": "LeaderElection", "message": "karpenter-65456c857f-ztqxq_28696d7c-242a-4920-9879-06dcef5332e4 became leader"}
  | Jun 20 17:56:38.814 | karpenter-65456c857f-ztqxq | successfully acquired lease karpenter/karpenter-leader-election
  | Jun 20 17:56:32.287 | karpenter-65456c857f-ct9dl | 2022-06-20T21:56:31.794Z DEBUG controller.aws.launchtemplate Finished hydrating the launch template cache with 1 items {"commit": "10fc37b"}
  | Jun 20 17:56:32.287 | karpenter-65456c857f-ct9dl | 2022-06-20T21:56:31.642Z DEBUG controller.aws.launchtemplate Hydrating the launch template cache with names matching "Karpenter-eks-us-west-2-*" {"commit": "10fc37b"}
  | Jun 20 17:56:31.824 | karpenter-65456c857f-ct9dl | attempting to acquire leader lease karpenter/karpenter-leader-election...
  | Jun 20 17:56:31.343 | karpenter-65456c857f-ztqxq | successfully acquired lease karpenter/webhook.webhookcertificates.00-of-01
  | Jun 20 17:56:30.730 | karpenter-65456c857f-ztqxq | successfully acquired lease karpenter/webhook.defaultingwebhook.00-of-01
  | Jun 20 17:56:30.719 | karpenter-65456c857f-ztqxq | attempting to acquire leader lease karpenter/webhook.defaultingwebhook.00-of-01...
  | Jun 20 17:56:30.719 | karpenter-65456c857f-ztqxq | failed to renew lease karpenter/webhook.defaultingwebhook.00-of-01: timed out waiting for the condition
  | Jun 20 17:56:30.555 | karpenter-65456c857f-ztqxq | attempting to acquire leader lease karpenter/webhook.webhookcertificates.00-of-01...
  | Jun 20 17:56:30.551 | karpenter-65456c857f-ztqxq | failed to renew lease karpenter/webhook.webhookcertificates.00-of-01: timed out waiting for the condition
  | Jun 20 17:56:30.532 | karpenter-65456c857f-ztqxq | successfully acquired lease karpenter/webhook.validationwebhook.00-of-01
  | Jun 20 17:56:30.520 | karpenter-65456c857f-ztqxq | attempting to acquire leader lease karpenter/webhook.validationwebhook.00-of-01...
  | Jun 20 17:56:30.520 | karpenter-65456c857f-ztqxq | failed to renew lease karpenter/webhook.validationwebhook.00-of-01: timed out waiting for the condition
  | Jun 20 17:56:30.222 | karpenter-65456c857f-ct9dl | failed to renew lease karpenter/karpenter-leader-election: timed out waiting for the condition
  | Jun 20 17:51:37.770 | karpenter-65456c857f-ct9dl | 2022-06-20T21:51:37.277Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 17:38:20.057 | karpenter-65456c857f-ct9dl | 2022-06-20T21:38:19.731Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 15:52:05.896 | karpenter-65456c857f-ct9dl | 2022-06-20T19:52:05.277Z DEBUG controller.aws.launchtemplate Deleted launch template lt-05cf56c6e691623ba {"commit": "10fc37b"}
  | Jun 20 15:48:44.899 | karpenter-65456c857f-ct9dl | 2022-06-20T19:48:44.096Z DEBUG controller Discovered launch template Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 15:45:08.900 | karpenter-65456c857f-ct9dl | 2022-06-20T19:45:08.357Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 15:32:05.668 | karpenter-65456c857f-ct9dl | 2022-06-20T19:32:05.288Z DEBUG controller.aws.launchtemplate Deleted launch template lt-02586e4c771e80dc2 {"commit": "10fc37b"}
  | Jun 20 15:20:39.666 | karpenter-65456c857f-ct9dl | 2022-06-20T19:20:38.915Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 14:52:05.665 | karpenter-65456c857f-ct9dl | 2022-06-20T18:52:05.267Z DEBUG controller.aws.launchtemplate Deleted launch template lt-0eb027493e13efa79 {"commit": "10fc37b"}
  | Jun 20 14:42:11.672 | karpenter-65456c857f-ct9dl | 2022-06-20T18:42:10.901Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 14:12:05.937 | karpenter-65456c857f-ct9dl | 2022-06-20T18:12:05.267Z DEBUG controller.aws.launchtemplate Deleted launch template lt-01d3c636fed2cb244 {"commit": "10fc37b"}
  | Jun 20 14:09:00.941 | karpenter-65456c857f-ct9dl | 2022-06-20T18:09:00.058Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 13:32:05.981 | karpenter-65456c857f-ct9dl | 2022-06-20T17:32:05.326Z DEBUG controller.aws.launchtemplate Deleted launch template lt-03876fea01961e530 {"commit": "10fc37b"}
  | Jun 20 13:21:59.656 | karpenter-65456c857f-ct9dl | 2022-06-20T17:21:59.525Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 13:12:05.643 | karpenter-65456c857f-ct9dl | 2022-06-20T17:12:05.256Z DEBUG controller.aws.launchtemplate Deleted launch template lt-0248e2690011b08ed {"commit": "10fc37b"}
  | Jun 20 13:03:06.648 | karpenter-65456c857f-ct9dl | 2022-06-20T17:03:06.190Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 12:32:05.287 | karpenter-65456c857f-ct9dl | 2022-06-20T16:32:05.272Z DEBUG controller.aws.launchtemplate Deleted launch template lt-07f2ded5df6d51aa7 {"commit": "10fc37b"}
  | Jun 20 12:27:29.291 | karpenter-65456c857f-ct9dl | 2022-06-20T16:27:28.708Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10fc37b", "provisioner": "default"}
  | Jun 20 11:42:05.787 | karpenter-65456c857f-ct9dl | 2022-06-20T15:42:05.261Z DEBUG controller.aws.launchtemplate Deleted launch template lt-07f0c0cddf53a5658 {"commit": "10fc37b"}
  | Jun 20 11:36:07.615 | karpenter-65456c857f-ct9dl | 2022-06-20T15:36:07.089Z DEBUG controller Created launch template, Karpenter-eks-us-west-2-7225536031179167781 {"commit": "10
```

From the logs above a wild guess: the lease expired and the leader rotated, however the old leader has still the in memory cache and when the cache evicted he tried to delete the launch template which he shouldn't

**Comment by dewjam** [comment]
Hey @nandiheath . 
Thanks for reporting this issue.  I'm going to try to reproduce the problem and debug.  I'll let you know what I find.

**Comment by spring1843** [comment]
We have internally confirmed the problem. We probably need to handle this situation correctly by using this channel properly during leader selection https://github.com/kubernetes-sigs/controller-runtime/blob/196828e54e4210497438671b2b449522c004db5c/pkg/manager/manager.go#L66

**Comment by nandiheath** [comment]
thank you for the info. 
Do you have a ETA when the fix will be released? we are running HA on our production clusters and would like to remove the custom launch template for enabling node TTL but this could potentially block us from adopting that

**Comment by dewjam** [comment]
Hello @nandiheath ,
This is planned to be included in the v0.13.0 release of Karpenter which is expected to land this week.
